### PR TITLE
#342 Implement "Bring your own key" feature

### PR DESCRIPTION
**Pull request by dsargent**
## Overview
BYOK (bring your own key) functionality for MAUI. Detects when the server tells us the network requires API keys for OpenAI or Anthropic (easy to extend for other providers), and sends the required keys if the user has entered them. Updates "Settings" dialog to allow users to enter and test their API keys.

## Screens
Users can enter their API keys in the "Settings" dialog:
<br />
image
<br />
Successful test:
image
<br />
When Neuro-san informs the UI that a key (or keys) is/are required for the current network, MAUI will send the requested keys if present. If the user hasn't provided keys, they will be blocked from interacting with the network:
<br />
image
<br />
## Known issues
* ⚠️ Keys are stored in plain text in browser local storage. This is not ideal as they are vulnerable to exfiltration by malicious extensions or XSS attacks. Encrypting them wouldn't help, as then the encryption key itself would have to reside in the browser and would be available to the same attacker. A better solution would be to store them server-side, encrypted, but that does not work with our current architectural standards (no state/no DB/no server-side storage).
* If the user inputs their key but doesn't save it, we don't pop up a confirmation when they close the dialog. This is not ideal, but requires more thought and design as the Settings dialog has traditionally been a "no confirm, just apply immediately" paradigm; keys work differently so that clashes. But I think this is okay for MVP level.
* For now, supports OpenAI and Anthropic, but extensible to arbitrary numbers of providers.

### CODE REVIEW

File: packages/ui-common/components/Settings/ApiKeyInput.tsx [hunk 1]
```diff
@@ -0,0 +1,174 @@
+import CheckIcon from "@mui/icons-material/Check"
+import ClearIcon from "@mui/icons-material/Clear"
+import ErrorIcon from "@mui/icons-material/Error"
+import Box from "@mui/material/Box"
+import Button from "@mui/material/Button"
+import FormLabel from "@mui/material/FormLabel"
+import IconButton from "@mui/material/IconButton"
+import InputAdornment from "@mui/material/InputAdornment"
+import TextField from "@mui/material/TextField"
+import {FC, ChangeEvent as ReactChangeEvent, useState} from "react"
```
[the remark this answers is not in the document]
Comment (dsargent) [reply packages/ui-common/components/Settings/ApiKeyInput.tsx]: Go home, Copilot you're drunk.

File: apps/main/public/claude.png
Comment (dsargent) [inline apps/main/public/claude.png:1]: Logo for Settings dialog

File: apps/main/public/OpenAI-black.png
Comment (dsargent) [inline apps/main/public/OpenAI-black.png:1]: Logo for Settings dialog

File: apps/main/public/OpenAI-white.png
Comment (dsargent) [inline apps/main/public/OpenAI-white.png:1]: Logo for Settings dialog (dark mode)

File: packages/dev-common/Configs/eslint.config.mjs [hunk 2]
```diff
@@ -124,6 +124,9 @@ export default defineConfig([
             // Disable by default - consumers must configure with their project-specific pages path
             "@next/next/no-html-link-for-pages": "off",
 
+            // We don't care about Next.js image optimization
```
Comment (dsargent) [inline packages/dev-common/Configs/eslint.config.mjs:127]: Tired of this rule. We already disable it everywhere, so I just global disabled it. We don't care about Next.js.

File: packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx [hunk 3]
```diff
@@ -84,8 +84,7 @@ describe("ChatCommon", () => {
         onSend: jest.fn(),
         sampleQueries: MOCK_CONNECTIVITY_INFO.metadata["sample_queries"],
         setIsAwaitingLlm: jest.fn(),
-        targetAgent: TEST_AGENT_MATH_GUY,
-        userImage: "",
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx:88]: Follow up from UI redesign PR: we no longer display user images (or bot images) in Chat.

File: packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx [hunk 4]
```diff
@@ -332,7 +331,10 @@ describe("ChatCommon", () => {
         const testResponseText1 = "Response text 1 from LLM"
         const testResponseText2 = "Response text 2 from LLM"
 
-        renderChatCommonComponent({onChunkReceived: onChunkReceivedMock, targetAgent: LegacyAgentType.DataGenerator})
+        renderChatCommonComponent({
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx:334]: I renamed `targetAgent` to `selectedNetwork` throughout. This matches our terminology elsewhere like in the main `MultiAgentAccelerator` component and reduces confusion. 

I think I originally named it "target agent" because of its dual role, supporting legacy agents which aren't networks but just individual "agents", as well as Neuro-san networks. But it's just confusing to have it named differently than what other components call it. Now it's consistent.

File: packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx [hunk 5]
```diff
@@ -353,7 +355,7 @@ describe("ChatCommon", () => {
         const finalAnswerText = "Sample final answer from LLM"
         const testResponseText = `Final Answer: ${finalAnswerText}`
 
-        renderChatCommonComponent({onChunkReceived: onChunkReceivedMock, targetAgent: LegacyAgentType.DMSChat})
+        renderChatCommonComponent({onChunkReceived: onChunkReceivedMock, selectedNetwork: LegacyAgentType.DMSChat})
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx:358]: Lots of tiny changes due to the rename.

File: packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx [hunk 6]
```diff
@@ -638,11 +640,28 @@ describe("ChatCommon", () => {
...
+        const overlay = document.getElementById("chat-disabled-overlay")
+        expect(overlay).toHaveStyle({
+            position: "absolute",
+            zIndex: MODAL_Z_INDEX - 1,
+            cursor: "not-allowed",
+            pointerEvents: "all",
+        })
+    })
+
+    it("Should refuse interaction when API keys are required but not present", async () => {
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx:657]: Diffs are wonky but this is a new test for the new "missing API keys" overlay.

File: packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx [hunk 7]
```diff
@@ -790,7 +809,22 @@ describe("ChatCommon", () => {
 
         // No final answer from a Neuro-san agent; this is an error
         const alertItem = screen.getByRole("alert")
-        within(alertItem).getByText(/final answer/u)
+        within(alertItem).getByText(/did not provide a final answer/u)
+    })
+
+    it("Should handle when Neuro-san agents fail to send a final answer", async () => {
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/AgentChat/ChatCommon/ChatCommon.test.tsx:815]: Testing another edge case (okay, just trying to up the coverage numbers.)

File: packages/ui-common/__tests__/components/MultiAgentAccelerator/MultiAgentAccelerator.test.tsx [hunk 8]
```diff
@@ -211,13 +211,13 @@ jest.mock("../../../components/AgentChat/ChatCommon/ChatCommon", () => ({
             <div
                 id="test-chat-common"
                 data-testid="test-chat-common"
-            />
+            >
+                {props.selectedNetwork}
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/MultiAgentAccelerator/MultiAgentAccelerator.test.tsx:215]: Makes the tests more faithful to the actual behavior.

File: packages/ui-common/__tests__/components/MultiAgentAccelerator/MultiAgentAccelerator.test.tsx [hunk 9]
```diff
@@ -1199,14 +1199,113 @@ describe("MultiAgentAccelerator", () => {
...
+                        selectedNetwork: AGENT_NETWORK_DESIGNER_ID,
                         extraSlyData: {
                             [AGENT_NETWORK_DEFINITION_KEY]: agentNetworkDefinition,
                         },
                     })
                 )
             })
         })
+
+        const badProvider = "not_a_known_provider"
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/MultiAgentAccelerator/MultiAgentAccelerator.test.tsx:1211]: Tests for the various combinations of "keys required" vs "keys that we have".

File: packages/ui-common/__tests__/components/MultiAgentAccelerator/MultiAgentAccelerator.test.tsx [hunk 10]
```diff
@@ -1199,14 +1199,113 @@ describe("MultiAgentAccelerator", () => {
...
+        })
+
+        it.each([
+            {},
+            {function: {}},
+            {function: {sly_data_schema: {}}},
+            {function: {sly_data_schema: {properties: {}}}},
+            {function: {sly_data_schema: {properties: {llm_config: {}}}}},
+            {function: {sly_data_schema: {properties: {llm_config: {required: []}}}}},
+        ])("handles missing items in the BYOK info from the backend", async (byokResponse) => {
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/MultiAgentAccelerator/MultiAgentAccelerator.test.tsx:1273]: Strange looking test to exercise the various gaps in the JSON that we can encounter from the server. This is the price of `sly_data` being "typed" as  `Record<string, unknown>`.

File: packages/ui-common/__tests__/components/Settings/SettingsDialog.test.tsx [hunk 11]
```diff
@@ -64,6 +67,115 @@ describe("SettingsDialog", () => {
         expect(onCloseMock).toHaveBeenCalledTimes(1)
     })
 
+    describe("API keys", () => {
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/Settings/SettingsDialog.test.tsx:70]: Test suite for new API keys section in Settings dialog.

File: packages/ui-common/__tests__/components/Settings/SettingsDialog.test.tsx [hunk 12]
```diff
@@ -132,6 +244,25 @@ describe("SettingsDialog", () => {
         expect(useSettingsStore.getState().settings.appearance.rangePalette).toBe("grayScale")
     })
 
+    it("doesn't change palette if user clicks again on the existing option", async () => {
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/Settings/SettingsDialog.test.tsx:281]: Filling in gaps in coverage.

File: packages/ui-common/__tests__/unit/Controller/Providers.test.ts
Comment (dsargent) [inline packages/ui-common/__tests__/unit/Controller/Providers.test.ts:1]: Tests for new controller.

File: packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx [hunk 13]
```diff
@@ -93,9 +88,9 @@ export interface ChatCommonProps {
     readonly isAwaitingLlm: boolean
 
     /**
-     * The agent to send the request to.
+     * The network to send the request to.
      */
-    readonly targetAgent: string
+    readonly selectedNetwork: string | null
```
Comment (dsargent) [inline packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx:93]: The renaming I mentioned earlier. Lot of diffs due to this.

File: packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx [hunk 14]
```diff
@@ -181,6 +176,11 @@ export interface ChatCommonProps {
      * Sample queries for the current network that the user can "click to send"
      */
     readonly sampleQueries?: string[]
+
+    /**
```
Comment (dsargent) [inline packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx:180]: This is solely so we can throw up the overlay and block interaction if we don't have the required keys.

File: packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx [hunk 15]
```diff
@@ -273,7 +274,10 @@ export const ChatCommon = ({ref, ...props}: ChatCommonProps & {ref?: Ref<ChatCom
 
     // Persistent agent chat history store, which is where we store both kinds of chat histories
     // (see store implementation for details)
-    const storedChatHistory = useAgentChatHistoryStore((state) => state?.history?.[targetAgent])
+    const storedChatHistory = useAgentChatHistoryStore((state) =>
```
Comment (dsargent) [inline packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx:277]: Copilot comment -- handle missing network name.

File: packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx [hunk 16]
```diff
@@ -645,25 +649,19 @@ export const ChatCommon = ({ref, ...props}: ChatCommonProps & {ref?: Ref<ChatCom
...
-
-            // Save final answer to chat history
-            const finalAnswerContent = finalAnswerTurn.text || JSON.stringify(finalAnswerTurn.structure, null, 2)
-            updateChatHistory(targetAgent, [new AIMessage({content: finalAnswerContent, id: uuid()})])
-        } else {
-            // No final answer found, display error
-            setTurns((prev) => [...prev, getFinalAnswerErrorTurn()])
-        }
-    }, [targetAgent, updateChatHistory])
+        // Update relevant turn to be the final answer
```
Comment (dsargent) [inline packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx:652]: I simplified the logic here. Firstly, it was super hard to hit the various cases with tests. Then I realized, that was because some of them are literally impossible; we check for "no text, no structure" when handling incoming chunks, so messages like that wouldn't even make it this far.

I've started realizing, with our super high test coverage, that "impossible to hit with tests" is sometimes a sign that a condition can actually never logically happen.

File: packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx [hunk 17]
```diff
@@ -781,26 +779,32 @@ export const ChatCommon = ({ref, ...props}: ChatCommonProps & {ref?: Ref<ChatCom
         [handleStop, handleClearChat]
     )
 
-    const getNoAgentOverlay = () => (
-        <Tooltip
-            title="Please select a Network from the list to start the chat."
-            placement="auto"
+    const getErrorOverlay = (errorText: string) => (
```
Comment (dsargent) [inline packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx:782]: Renamed this because it now has different purposes (API keys missing or no network selected). Could also be factored to its own component.

I considered MUI `Backdrop` here but it doesn't really do what we want.

File: packages/ui-common/components/ChatBot/ChatBot.tsx [hunk 18]
```diff
@@ -33,11 +33,6 @@ interface ChatBotProps {
      */
     readonly id: string
 
-    /**
```
Comment (dsargent) [inline packages/ui-common/components/ChatBot/ChatBot.tsx:36]: No longer displayed in Chat.

File: packages/ui-common/components/Common/CustomerLogo.tsx [hunk 19]
```diff
@@ -24,7 +24,6 @@ export const getCognizantLogoImage = () => (
         target="_blank"
         rel="noopener noreferrer"
     >
-        {/* eslint-disable-next-line @next/next/no-img-element */}
```
Comment (dsargent) [inline packages/ui-common/components/Common/CustomerLogo.tsx:27]: Now disabled globally.

File: packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx [hunk 20]
```diff
@@ -289,25 +294,79 @@ export const MultiAgentAccelerator: FC<MultiAgentAcceleratorProps> = ({
...
+        }
+
+        return {llm_config: llmConfig}
+    }
+
+    /**
+     * Builds the extraSlyData object to be sent with each request, including information for Agent Network Designer
+     * and (if required) API keys for LLM providers.
+     */
+    const buildExtraSlyData = (): Record<string, unknown> | undefined => {
```
Comment (dsargent) [inline packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx:319]: I modified the existing "extra sly data" logic that you added @swensel to inject the API keys when (1) we have them and (2) the network metadata shows they are required.

File: packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx [hunk 21]
```diff
@@ -333,8 +392,32 @@ export const MultiAgentAccelerator: FC<MultiAgentAcceleratorProps> = ({
             try {
                 const agentFunction = await getAgentFunction(neuroSanURL, selectedNetwork, userInfo.userName)
                 setNetworkDescription(agentFunction?.function?.description || "")
-            } catch {
-                // Ignore. May be a legacy agent without a functional description in Neuro-san.
+
+                const schema = agentFunction?.function?.sly_data_schema
```
Comment (dsargent) [inline packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx:396]: "Dealing with the `unknown`". The schema of `sly_data` guarantees _nothing_ so we have to tread gingerly.

File: packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx [hunk 22]
```diff
@@ -667,6 +750,10 @@ export const MultiAgentAccelerator: FC<MultiAgentAcceleratorProps> = ({
         }
     }, [tourRequested, selectedNetwork, agentsInNetwork, networks, controls, setTourStatus])
 
+    const getMissingApiKeys = () => {
+        return providerKeysRequired.size > 0 ? [...providerKeysRequired].filter((provider) => !apiKeys?.[provider]) : []
```
Comment (dsargent) [inline packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx:754]: Once we move to a newer ES target like `es2024` (#395), there's a whole array of set operations like intersection, set-diff that can do this safer and more readably. But for now we're stuck with the hand-crafted way.

File: packages/ui-common/components/Settings/ApiKeyInput.tsx
Comment (dsargent) [inline packages/ui-common/components/Settings/ApiKeyInput.tsx:1]: New component to accept, validate and persist API keys. Hopefully reusable across other apps.

I did add some "quality of life" frills:
1) "Right to be forgotten" - can clear out the saved keys
2) Ability to test keys -- should be very handy
3) Show/hide keys to look for typos/copy-paste issues

File: packages/ui-common/__tests__/components/Settings/SettingsDialog.test.tsx [hunk 23]
```diff
@@ -132,6 +278,25 @@ describe("SettingsDialog", () => {
...
```
Comment (dsargent) [inline packages/ui-common/__tests__/components/Settings/SettingsDialog.test.tsx:281]: New test to up coverage.

File: packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx [hunk 24]
```diff
@@ -1146,7 +1150,14 @@ export const ChatCommon = ({ref, ...props}: ChatCommonProps & {ref?: Ref<ChatCom
                 position: "relative",
             }}
         >
-            {targetAgent ? getChatBox() : getNoAgentOverlay()}
+            {selectedNetwork
```
Comment (dsargent) [inline packages/ui-common/components/AgentChat/ChatCommon/ChatCommon.tsx:1153]: Display new overlay if keys are missing.

File: packages/ui-common/components/ChatBot/ChatBot.tsx [hunk 25]
```diff
@@ -93,8 +88,7 @@ export const ChatBot: FC<ChatBotProps> = ({id, userAvatar, pageContext}) => {
                         currentUser={currentUser}
                         setIsAwaitingLlm={setIsAwaitingLlm}
                         isAwaitingLlm={isAwaitingLlm}
-                        targetAgent={LegacyAgentType.ChatBot}
```
Comment (dsargent) [inline packages/ui-common/components/ChatBot/ChatBot.tsx:96]: Rename `targetAgent` + remove `userImage`.

File: packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx [hunk 26]
```diff
@@ -289,25 +294,79 @@ export const MultiAgentAccelerator: FC<MultiAgentAcceleratorProps> = ({
...
-              ...(currentTempNetwork.networkHocon ? {[AGENT_NETWORK_HOCON]: currentTempNetwork.networkHocon} : {}),
-          }
-        : designerTempNetwork
-          ? {[AGENT_NETWORK_DEFINITION_KEY]: designerTempNetwork.agentNetworkDefinition}
-          : undefined
+    /**
+     * Builds the API keys object to be sent as extraSlyData with each request, based on the LLM providers required
+     * by the currently selected network. Only includes keys for providers that are required.
+     */
+    const getApiKeys = () => {
```
Comment (dsargent) [inline packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx:301]: Would be nice to have this more data-driven but we're dealing with schema-less strings from the API (not enums or anything predictable).

File: packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx [hunk 27]
```diff
@@ -772,18 +859,19 @@ export const MultiAgentAccelerator: FC<MultiAgentAcceleratorProps> = ({
...
                     <ChatCommon
-                        customAgentGreetings={{
-                            [AGENT_NETWORK_DESIGNER_ID]: "Let's build a network together!",
-                        }}
                         agentPlaceholders={{
                             [AGENT_NETWORK_DESIGNER_ID]:
                                 "Describe in plain language the network you would like to build.",
                         }}
                         currentUser={userInfo.userName}
+                        customAgentGreetings={{
```
Comment (dsargent) [inline packages/ui-common/components/MultiAgentAccelerator/MultiAgentAccelerator.tsx:867]: Sort alphabetically

File: packages/ui-common/components/Settings/SettingsDialog.tsx [hunk 28]
```diff
@@ -233,6 +261,29 @@ export const SettingsDialog: FC<SettingsDialogProps> = ({id, isOpen, logoService
                     border: "1px solid",
                 }}
             >
+                <Box sx={{marginBottom: 3}}>
```
Comment (dsargent) [inline packages/ui-common/components/Settings/SettingsDialog.tsx:264]: New dialog section for keys

File: packages/ui-common/controller/llm/Providers.ts
Comment (dsargent) [inline packages/ui-common/controller/llm/Providers.ts:1]: New controller for testing LLM provider keys. Couldn't think of a better name. Didn't want something too specific and clunky like "KeyTester".

File: packages/ui-common/state/Settings.ts [hunk 29]
```diff
@@ -53,6 +55,7 @@ interface Settings {
     readonly behavior: {
         readonly enableZenMode: boolean
     }
+    readonly apiKeys: Partial<Record<LLMProvider, string>>
```
Comment (dsargent) [inline packages/ui-common/state/Settings.ts:58]: New persistence field for keys.

File: jest.config.ts [hunk 30]
```diff
@@ -94,10 +94,10 @@ const config: Config.InitialOptions = {
 
     coverageThreshold: {
         global: {
-            statements: -126,
-            branches: -168,
-            functions: -33,
-            lines: -97,
+            statements: -124,
```
Comment (dsargent) [inline jest.config.ts:97]: Getting harder and harder to find improvements (which is good, it means we have high test coverage) but I scrounged some here.

**Review by swensel — APPROVED** [review]
Looks great. Good work @dsargent . 🎉
